src/dtree.ml

```diff
@@ -0,0 +1,192 @@
+        F.fprintf ppf "@ %d -- %d [label=\"%s\"];" father_l
+          !nodecount "l";
+      end
+    | Node(ndata) ->
+      let { switch = t ; swap = _ ; children = c } = ndata in
+      begin
+        incr nodecount ;
+        let tag = !nodecount in
+        F.fprintf ppf "@ %d -- %d [label=\"" father_l tag ;
+        begin
+          match t with
+          | Some t' -> Print.pp ppf t'
+          | None    -> F.fprintf ppf "d"
+        end ;
+        F.fprintf ppf "\"];" ;
+        List.iter (fun e -> write_tree tag e) c ;
+      end
+    | Fail        ->
+      begin
+        incr nodecount ;
+        F.fprintf ppf "@ %d -- %d [label=\"%s\"];" father_l
+          !nodecount "f";
+        end
+  in
+  write_tree 0 tree ;
+  F.fprintf ppf "@.}@\n@?" ;
+  close_out ochan
+
+module Pattmat =
+struct
+  (** Type used to describe a line of a matrix (either a column or a row). *)
+  type line = term list
+
+  (** Mainly for debugging purposes, shows where the matrix comes from,
+      i.e. generated by a specialization, a default case or the initial
+      matrix *)
+  type operation = Init
+                 | Default
+                 | Specialized of term
+
+  (** Contains the rewrite rules. *)
+  type matrix = (line * action) list
+
+  (** Type of a matrix of patterns.  Each line is a row having an attached
+      action. *)
+  type t = { origin : operation
+           ; values : matrix }
+
+  (** [pp_line l] prints line [l] to stdout. *)
+  let pp_line : line -> unit = fun l ->
+    let module F = Format in
+    let rec loop : term list -> unit = function
+      | []      -> ()
+      | x :: xs ->
+        begin
+          F.print_space () ; F.print_string ";" ;
+          Print.pp Format.std_formatter x ;
+          loop xs
+        end in
+    F.open_box 0 ;
+    F.print_string "[" ; loop l ; F.print_string  "]" ;
+    F.close_box () ; F.print_newline ()
+
+  (** [pp m] prints matrix [m] to stdout. *)
+  let pp : t -> unit = fun m ->
+    let module F = Format in
+    let rec loop : matrix -> unit = function
+      | []           -> ()
+      | (l, _) :: xs ->
+        begin
+          F.print_tab () ; pp_line l ; loop xs
+        end in
+    F.open_box 0 ;
+    F.print_string "[|" ; F.print_cut () ; loop m.values ; F.print_string "|]" ;
+    F.close_box () ; F.print_newline ()
+
+  (** [of_rules r] creates the initial pattern matrix from a list of rewriting
+      rules. *)
+  let of_rules : rule list -> t = fun rs ->
+    { values = List.map (fun r -> r.lhs, r.rhs) rs
+    ; origin = Init }
+
+  (** [is_empty m] returns whether matrix [m] is empty. *)
+  let is_empty : t -> bool = fun m -> List.length m.values = 0
+
+  (** [get_col n m] retrieves column [n] of matrix [m].  There is some
+      processing because all rows do not have the same length. *)
+  let get_col : int -> t -> line = fun ind mat ->
+    List.fold_left (fun acc (elt, _) ->
+        if List.length elt > ind then List.nth elt ind :: acc else acc) []
```

Comment:
This code goes through elt twice. Can't we avoid this and go through elt only once?

Reply:
Fixed in 87fddd4